dll/win32/kernel32/client/session.c

```diff
@@ -13,31 +16,141 @@
+            if (SessionID == 0)
+            {
+                OutPathLength += wcslen(L"GLOBALROOT\\DosDevices\\");
+                OutPathLength += InPathLength;
+            }
+            else
+            {
+                OutPathLength += wcslen(L"GLOBALROOT\\Sessions\\");
+                OutPathLength += 10; // Length of SessionID(char)
+                OutPathLength += wcslen(L"\\DosDevices\\");
+                OutPathLength += InPathLength;
+            }
+        }
+        else
+        {
+            OutPathLength += InPathLength;
+        }
+        OutPathLength += 1; // \0 termination
+        SessionPath = (LPWSTR)LocalAlloc(LMEM_FIXED, OutPathLength * sizeof(WCHAR));
+        if (SessionPath != 0)
+        {
+            if (BaseStaticServerData->LUIDDeviceMapsEnabled == TRUE)
+            {
+                RtlStringCchPrintfW(SessionPath, OutPathLength, L"%ws", InPath);
+            }
+            else if (SessionID == 0)
+            {  // GLOBALROOT\DosDevices\$InPath
+                RtlStringCchPrintfW(SessionPath, OutPathLength, L"GLOBALROOT\\DosDevices\\%ws", InPath);
+            }
+            else
+            { // GLOBALROOT\Sessions\$SessionID\DosDevices\$InPath
+                RtlStringCchPrintfW(SessionPath, OutPathLength, L"GLOBALROOT\\Sessions\\%u\\DosDevices\\%ws", SessionID, InPath);
+            }
+            *OutPath = SessionPath;
+            return 1;
+        }
+        else
+        {
+            LastError = ERROR_NOT_ENOUGH_MEMORY;
+        }
+    }
+    RtlSetLastWin32Error(LastError);
     return 0;
 }
 
 /*
- * @unimplemented
+ * @implemented
  */
 DWORD
 WINAPI
-DosPathToSessionPathA(IN DWORD SessionId,
+DosPathToSessionPathA(IN DWORD SessionID,
                       IN LPSTR InPath,
                       OUT LPSTR *OutPath)
 {
-    UNIMPLEMENTED;
+    DWORD LastError = 0;
+    if (InPath == NULL)
+    {
+        LastError = ERROR_INVALID_PARAMETER;
+    }
+    else
+    { // AnsiPath: InPath(ANSI) -> UnicodePath: InPath(Unicode) -> wcOutPath: OutPath(wide-char) -> UnicodePath: OutPath(Unicode) -> AnsiPath: OutPath(ANSI)
+        STRING AnsiPath = { 0 };
+        RtlInitAnsiString(&AnsiPath, InPath);
+        UNICODE_STRING UnicodePath;
+        NTSTATUS status = RtlAnsiStringToUnicodeString(&UnicodePath, &AnsiPath, TRUE); // UnicodePath = InPath(Unicode)
+        if (status < 0)
+        {
+            BaseSetLastNTError(status);
+            return 0;
+        }
+        LPWSTR wcOutPath = NULL; // OutPath(wide-char)
+        if (!DosPathToSessionPathW(SessionID, UnicodePath.Buffer, &wcOutPath))
+        {
+            RtlFreeUnicodeString(&UnicodePath);
+            return 0;
+        }
+        RtlFreeUnicodeString(&UnicodePath);
+        RtlInitUnicodeString(&UnicodePath, wcOutPath); // UnicodePath = OutPath(Unicode)
+        status = RtlUnicodeStringToAnsiString(&AnsiPath, &UnicodePath, TRUE);
+        if (status < 0)
```

Comment:
```suggestion
        if (!NT_SUCCESS(status))
```
Use `NT_SUCCESS` macro to check `NTSTATUS`.